apps/omg_child_chain/lib/omg_child_chain/event_fetcher.ex

```diff
@@ -0,0 +1,269 @@
+    # empty of logs.
+    # for the whole from, to range and signatures we create group pairs like so:
+    # from = 286, to = 287 signatures = ["Exit", "Deposit"]
+    # [{286, "Exit", []},{286, "Deposit", []},{287, "Exit", []},{287, "Deposit", []}]
+    empty_blknum_signature_events =
+      from_block..to_block
+      |> Enum.to_list()
+      |> Enum.map(fn blknum -> Enum.map(event_signatures, fn signature -> {blknum, signature, []} end) end)
+      |> List.flatten()
+
+    # we now merge the two lists
+    # it is important that logs we got from RPC are first
+    # because uniq_by takes the first occurance of {blknum, signature}
+    # so that we don't overwrite retrieved logs
+    data =
+      decoded_logs_in_keypair
+      |> Enum.concat(empty_blknum_signature_events)
+      |> Enum.uniq_by(fn {blknum, signature, _data} ->
+        {blknum, signature}
+      end)
+
+    :ets.insert(state.ets_bucket, data)
+  end
+
+  # delete everything older then (current block - delete_events_treshold)
+  defp delete_old_logs(new_height_blknum, state) do
+    # :ets.fun2ms(fn {block_number, _event_signature, _event} when
+    # block_number <= new_height - delete_events_treshold -> true end)
+    match_spec = [
+      {{:"$1", :"$2", :"$3"},
+       [{:"=<", :"$1", {:-, {:const, new_height_blknum}, {:const, state.delete_events_treshold_height_blknum}}}],
+       [true]}
+    ]
+
+    :ets.select_delete(state.ets_bucket, match_spec)
+  end
+
+  # allow ethereum event listeners to retrieve logs from ETS in bulk
+  defp handout_log(signature, from_block, to_block, state) do
+    # :ets.fun2ms(fn {block_number, event_signature, event} when
+    # block_number >= from_block and block_number <= to_block
+    # and event_signature == signature -> event and
+    # end)
+    event_match_spec = [
+      {{:"$1", :"$2", :"$3"},
+       [
+         {:andalso, {:andalso, {:>=, :"$1", {:const, from_block}}, {:"=<", :"$1", {:const, to_block}}},
+          {:==, :"$2", {:const, signature}}}
+       ], [:"$3"]}
+    ]
+
+    block_range = [
+      {{:"$1", :"$2", :"$3"},
+       [
+         {:andalso, {:andalso, {:>=, :"$1", {:const, from_block}}, {:"=<", :"$1", {:const, to_block}}},
+          {:==, :"$2", {:const, signature}}}
+       ], [:"$1"]}
+    ]
+
+    events = state.ets_bucket |> :ets.select(event_match_spec) |> List.flatten()
+    blknum_list = :ets.select(state.ets_bucket, block_range)
+
+    # we may not have all the block information the ethereum event listener wants
+    # so we check for that and find all logs for missing blocks
+    # in one RPC call for all signatures
+    case Enum.to_list(from_block..to_block) -- blknum_list do
+      [] ->
+        events
+
+      missing_blocks ->
+        missing_blocks = Enum.sort(missing_blocks)
+        missing_from_block = List.first(missing_blocks)
+        missing_to_block = List.last(missing_blocks)
+
+        _ =
```

Comment:
this isn't so much cache - maybe I need to rephrase the whole approach...

this happens as often as ethereum event listeners move across time. the *first* process reaching forward would hit this clause because that's when the EventFetcher would gather the events from contracts (but NOT before that)

so a single process hitting this code would not necessarily mean that additional RPC calls are needed, but merely that you're moving forward and Event Fetcher really hasn't been there.

On the other hand, as I'm moving forward, I'm also deleting past events (search for `delete_events_treshold_height_blknum `). if this wouldn't be large enough -than slower processes - those that have a bigger margin) could be trying to get stuff that was already deleted